### DISCUSSION TITLE
Custom ruleset score servers

### DISCUSSION THREAD

**Discussion by orangixx** [ppy/osu#13822] [Ideas] [answered]
Rather than blocking submission (#13819/#13820), custom rulesets could be able to use their own [osu-web](/ppy/osu-web) servers. (Of course, optional for both the developer and player.)

#11736 _could_ supersede this, but I personally believe in more decentralization.

**Comment by bdach** [comment]
Was suggested before. See #11592.

Also note that hosting "own osu-web" instances is something not to be done without a lot of care (by which I mean ensuring no infringement of either the web AGPL license or the osu! trademark, and ensuring that the end result could not be potentially mistaken for the real production osu-web instance).

**Reply by orangixx** [reply]
> Was suggested before. See #11592.

#11592 is completely different - it's a decentralized data store proposal. The implementation I was thinking of was to have each ruleset expose its own server as a value in its top namespace:

```cs
namespace osu.Game.Rulesets.Foo {
    //...
    public const string SCORE_SERVER = "foo.bar.xyz";
    // ...
}
```

"" or not existing means no server.

If it were set to "osu.ppy.sh" it would obviously respond with an error. If #11736 is implemented then that would change. But I wish for more decentralization.

---

> Also note that hosting "own osu-web" instances is something not to be done without a lot of care (by which I mean ensuring no infringement of either the web AGPL license&hellip;

Most (if not all) rulesets are currently open source, and the developers wouldn't have a reason not to open source a modified site.

> &hellip;or the osu! trademark, and ensuring that the end result could not be potentially mistaken for the real production osu-web instance).

Phishing is unlikely to happen as a _result_ of this - all of the site is open source already.

Further customization is not difficult to implement.
I haven't combed through the code, but I disabled JavaScript on the site and it seems to use server-side rendering for the base of the page (CSS and header/footer) and much of the front page, which should not be too hard to replace. A lot of the client-side code might contain references to _osu!_ which would be harder to configure (osu-web could include an extra API endpoint for requesting the site name).

Also such servers would not _necessarily_ have to use osu-web: [Hanayo](https://zxq.co/ripple/hanayo) and similar softwares exist.

**Comment by peppy** [comment, marked as answer]
this is not a direction i am interested in supporting. please make your own stand alone game if you want to run the infrastructure.
